### DISCUSSION TITLE
`'abaqus' is not a package` - Trying to avoid `from abaqus import *`

### DISCUSSION THREAD

**Discussion by sebastianvitterso** [haiiliin/abqpy#5651] [Q&A] [answered]
Hey!

I'm currently trying to refactor a project that uses Abaqus to make it more developer friendly, but I'm getting an issue: 

Your examples, and really every other example in the Abaqus ecosystem, uses
```python
from abaqus import *
from abaqusConstants import *
from caeModules import *
```
at the top of almost every file.

I'd prefer to use named imports (as *-imports aren't the coolest to work with, software-dev-wise), but I'm seeing a hiccup when I'm trying to run the script (in Abaqus):

```python
ImportError: No module named 'abaqus.Model'; 'abaqus' is not a package
#* "/project/de4rm/habit/venv/lib/python3.11/site-packages/habit_to_abaqus/sketch.py",
#* line 4, in <module>
#*     from abaqus.Model.Model import Model
```

My autocomplete (in vscode) tells me this should work, and pylance/pyright certainly agrees.

Do you have any idea if there is some way to make this work?

Kind regards.

**Comment by haiiliin** [comment]
Could you provide more details about how you run the script? I am not able to determine why this happens, abaqus should be a package in abqpy.

**Reply by sebastianvitterso** [reply]
Oh yeah, this isn't a problem with abqpy in that sense. The problem arises when I'm running this inside Abaqus CAE. 
The command we run is `/path/to/abaqus-2024 cae noGUI=/project/de4rm/habit/venv/lib/python3.11/site-packages/habit_to_abaqus/__main__.py -- --custom-script-arguments`, which triggers the code, and in turn triggers that error.

**Reply by haiiliin** [reply]
You should not import subpackages of abaqus in your script because it does not exist in the abaqus python interpreter.

**Reply by sebastianvitterso** [reply]
Alright, so everything has to be imported from `abaqus`, then? Every example I find uses `from abaqus import *`, but I'd love to avoid that as long as possible, but if that's not doable, I guess I'll have to conform.

**Reply by haiiliin** [reply, marked as answer]
I don't know anything about where the class `Model` is located in abaqus, but maybe you can use something like this:

```python
from typing import TYPE_CHECKING, Type
from abaqus import mdb

if TYPE_CHECKING:
    from abaqus.Model.Model import Model as _Model

Model: Type["_Model"] = type(mdb.models["Model-1"])
```

Of course this is only applied on abaqus 2024 with python 3.10

**Reply by sebastianvitterso** [reply]
That's a good idea, I'll keep that in mind. And I guess I can import directly from abaqus (without *) still, just not use subimports. I tested, and it works without the `import *` as well. 

Thanks for quick feedback!

**Reply by diogo-rossi** [reply]
Hello guys.

Another way to get the same approach that @haiiliin [used here](https://github.com/haiiliin/abqpy/discussions/5651#discussioncomment-10007596) in older versions of Abaqus (<2024) is to check the Python version:

```python
import sys
from abaqus import mdb

if sys.version_info[0] >= 3: # Abaqus < 2024 uses python 2.7
    from abaqus.Model.Model import Model as _Model

Model = mdb.models["Model-1"] # type: _Model
```

Using this approach you can use type hints in the format **supported by Python 2.7**, which does not have `typing` module.

Additionally, your type checker must support this kind of type hints also.

**Reply by diogo-rossi** [reply]
Merging the approaches, the script should work with both versions:

```python
import sys
from abaqus import mdb

if sys.version_info[0] >= 3: # Abaqus < 2024 uses python 2.7
    from typing import TYPE_CHECKING 
    if TYPE_CHECKING: # Used only when developing with abqpy
        from abaqus.Model.Model import Model as _Model

Model = mdb.models["Model-1"] # type: _Model
```
